File: android/src/main/java/com/resideo/auth0_flutter/JSONHelpers.java
package com.resideo.auth0_flutter;

import com.auth0.android.authentication.AuthenticationException;
import com.auth0.android.authentication.storage.CredentialsManagerException;
import com.auth0.android.management.ManagementException;
import com.auth0.android.result.Credentials;
import com.auth0.android.result.UserIdentity;
import com.auth0.android.result.UserProfile;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class JSONHelpers {
    static HashMap<String, Object> credentialsToJSON(Credentials credentials) {
        HashMap<String, Object>  obj = new HashMap<>();
        obj.put("access_token", credentials.getAccessToken());
        obj.put("token_type", credentials.getType());
        obj.put("expires_in", credentials.getExpiresIn());
        obj.put("refresh_token", credentials.getRefreshToken());
        obj.put("id_token", credentials.getIdToken());
        obj.put("scope", credentials.getScope());

        return obj;
    }

    static HashMap<String, Object> profileTOJSON(UserProfile profile) {
        HashMap<String, Object> obj = new HashMap<>();
        obj.put("id", profile.getId());
        obj.put("sub", profile.getId());
        obj.put("name", profile.getName());
        obj.put("givenName", profile.getGivenName());
        obj.put("familyName", profile.getFamilyName());
        obj.put("nickname", profile.getNickname());

        obj.put("pictureURL", profile.getPictureURL());

        Date createdAt = profile.getCreatedAt();
        if (createdAt != null) {
            long milli = createdAt.getTime();
            long secs = milli / 1000;

            obj.put("createdAt", secs);
        }

        obj.put("email", profile.getEmail());
        obj.put("emailVerified", profile.isEmailVerified());

        final Map<String, Object> extraInfo = profile.getExtraInfo();
        extraInfo.put("user_metadata", profile.getUserMetadata());
        extraInfo.put("app_metadata", profile.getAppMetadata());

        obj.put("additionalAttributes", extraInfo);
        obj.put("user_metadata", profile.getUserMetadata());
        obj.put("app_metadata", profile.getAppMetadata());

        final List<UserIdentity> identities = profile.getIdentities();
        obj.put("identities", identitiesToJSON(identities));

        return obj;
    }

    static ArrayList<HashMap<String, Object>> identitiesToJSON(List<UserIdentity> identities) {
        if (identities == null) {
            return new ArrayList<>();
        }

        final ArrayList<HashMap<String, Object>> identitiesJSON = new ArrayList<>();

        for (UserIdentity identity : identities) {
            identitiesJSON.add(identityToJSON(identity));
        }

        return identitiesJSON;
    }

    static HashMap<String, Object> identityToJSON(UserIdentity identity) {
        HashMap<String, Object> obj = new HashMap<>();
        obj.put("identifier", identity.getId());
        obj.put("provider", identity.getProvider());
        obj.put("connection", identity.getConnection());
        obj.put("social", identity.isSocial());
        obj.put("profileData", identity.getProfileInfo());
        obj.put("accessToken", identity.getAccessToken());
        obj.put("accessTokenSecret", identity.getAccessTokenSecret());

        return obj;
    }

    static HashMap<String, Object> authErrorToJSON(AuthenticationException exception) {
        HashMap<String, Object> info = new HashMap<>();
        // Android Only
        info.put("isNetworkError", exception.isNetworkError());
        info.put("isBrowserAppNotAvailable", exception.isBrowserAppNotAvailable());
        info.put("isInvalidAuthorizeURL", exception.isInvalidAuthorizeURL());
        info.put("isInvalidConfiguration", exception.isInvalidConfiguration());
        info.put("isAuthenticationCanceled", exception.isAuthenticationCanceled());
        info.put("isPasswordLeaked", exception.isPasswordLeaked());
        // Cross platform
        info.put("isMultifactorRequired", exception.isMultifactorRequired());
        info.put("isMultifactorEnrollRequired", exception.isMultifactorEnrollRequired());
        info.put("isMultifactorTokenInvalid", exception.isMultifactorTokenInvalid());
        info.put("isMultifactorCodeInvalid", exception.isMultifactorCodeInvalid());
        info.put("isPasswordNotStrongEnough", exception.isPasswordNotStrongEnough());
        info.put("isPasswordAlreadyUsed", exception.isPasswordAlreadyUsed());
        info.put("isRuleError", exception.isRuleError());
        info.put("isInvalidCredentials", exception.isInvalidCredentials());
        // Web based
        info.put("isAccessDenied", exception.isAccessDenied());
        info.put("isLoginRequired", exception.isLoginRequired());

        HashMap<String, Object> obj = new HashMap<>();
        obj.put("code", exception.getCode());
        obj.put("statusCode", exception.getStatusCode());
        obj.put("description", exception.getDescription());
        obj.put("info", info);

        if (exception.isAuthenticationCanceled()) {
            obj.put("type", "userCancelled");
        } else {
            obj.put("type", "unknownError");
        }

        return obj;
    }

    static HashMap<String, String> credentialsErrorToJSON(CredentialsManagerException exception) {
        HashMap<String, String> obj = new HashMap<>();

        // TODO: unfortunate we have to check strings to determine the type of error returned.
        final String message = exception.getMessage();
        switch (message) {
            case "No Credentials were previously set.":
                obj.put("error_type", "no_credentials");
                break;
            case "Credentials have expired and no Refresh Token was available to renew them.":
                obj.put("error_type", "no_refresh_token");
                break;
            case "An error occurred while trying to use the Refresh Token to renew the Credentials.":
                obj.put("error_type", "failed_refresh");
                break;
            default:
                obj.put("error_type", "unknown");
                break;
        }

        obj.put("error_description", exception.getLocalizedMessage());

        return obj;
    }

    static HashMap<String, String> managementExceptionToJSON(ManagementException exception) {
        HashMap<String, String> obj = new HashMap<>();
        obj.put("code", exception.getCode());
        obj.put("description", exception.getDescription());

        return obj;
    }
}

File: ios/Classes/Helpers/JSONHelpers.swift
//
//  JSONHelpers.swift
//  auth0_flutter
//
//  Created by Juan Alvarez on 9/25/19.
//

import Foundation
import Auth0

extension Decodable {
    static func decode(data: Data, decoder: JSONDecoder = JSONDecoder()) throws -> Self {
        return try decoder.decode(Self.self, from: data)
    }
}

extension Credentials {
    func toJSON() -> [String: Any] {
        let json: [String : Any?] = [
            "access_token": self.accessToken,
            "token_type": self.tokenType,
            "expires_in": self.expiresIn?.timeIntervalSinceNow,
            "refresh_token": self.refreshToken,
            "id_token": self.idToken,
            "scope": self.scope
        ]
        
        let result = json.compactMapValues { $0 }
        
        return result
    }
}

func databaseUserToJSON(_ user: DatabaseUser) -> [String: Any] {
    let json: [String: Any?] = [
        "email": user.email,
        "username": user.username,
        "verified": user.verified
    ]
    
    let result = json.compactMapValues { $0 }
    
    return result
}

extension Identity {
    func toJSON() -> [String: Any] {
        let json: [String: Any?] = [
            "identifier": self.identifier,
            "provider": self.provider,
            "connection": self.connection,
            "social": self.social,
            "profileData": self.profileData,
            "accessToken": self.accessToken,
            "expiresIn": self.expiresIn?.timeIntervalSince1970,
            "accessTokenSecret": self.accessTokenSecret
        ]
        
        let result = json.compactMapValues { $0 }
        
        return result
    }
}

extension Profile {
    func toJSON() -> [String: Any] {
        let json: [String: Any?] = [
            "id": self.id,
            "name": self.name,
            "nickname": self.nickname,
            "pictureURL": self.pictureURL.absoluteString,
            "createdAt": self.createdAt.timeIntervalSince1970,
            "email": self.email,
            "emailVerified": self.emailVerified,
            "givenName": self.givenName,
            "familyName": self.familyName,
            "additionalAttributes": self.additionalAttributes,
            "identities": self.identities.map { $0.toJSON() },
        ]
        
        let result = json.compactMapValues { $0 }
        
        return result
    }
}

extension UserInfo {
    func toJSON() -> [String: Any] {
        let json: [String: Any?] = [
            "sub": self.sub,
            "name": self.name,
            "givenName": self.givenName,
            "familyName": self.familyName,
            "middleName": self.middleName,
            "nickname": self.nickname,
            "preferredUsername": self.preferredUsername,
            "profile": self.profile?.absoluteString,
            "picture": self.picture?.absoluteString,
            "website": self.website?.absoluteString,
            "email": self.email,
            "emailVerified": self.emailVerified,
            "gender": self.gender,
            "birthdate": self.birthdate,
            "zoneinfo": self.zoneinfo?.identifier,
            "locale": self.locale?.identifier,
            "phoneNumber": self.phoneNumber,
            "phoneNumberVerified": self.phoneNumberVerified,
            "address": self.address,
            "updatedAt": self.updatedAt?.timeIntervalSince1970,
            "customClaims": self.customClaims
        ]
        
        let result = json.compactMapValues { $0 }
        
        return result
    }
}
